### PR TITLE
[feat] Support using offset index in ParquetRecordBatchStream when pu…

### LINKED ISSUE #3456: Support using offset index in `ParquetRecordBatchStream` when pushing down `RowFilter`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

With https://github.com/apache/arrow-rs/pull/2552/files, using the offset index to prune page IO in `ParquetRecordBatchStream` should be stabilized. Once this is release in `arrow-rs` 23.0.0 we should allow users to enable offset indexes in `ParquetScanOptions`

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Ted-Jiang**
…shing down RowFilter.

Signed-off-by: yangjiang <yangjiang@ebay.com>

# Which issue does this PR close?

Closes #3456 

 # Rationale for this change
enable read `page index` in `ParquetScanOptions `
 If true, the reader will read pageIndex, If exit:
1. First we can use it create the `RowSelector` before read the file (like row-group pruning avoid I/O)
2. Second with pageIndex it will accelerate skip records (avoid decode pageHeader) when reading values from chunk with `RowSelector`.

# What changes are included in this PR?
update submodule `testing` and `parquet-testing`

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 1]
```diff
@@ -409,14 +411,31 @@ pub async fn fetch_parquet_metadata(
...
         let metadata_start = meta.size - length - 8;
 
-        Ok(decode_metadata(
-            &suffix[metadata_start - footer_start..suffix_len - 8],
-        )?)
+        decode_metadata(&suffix[metadata_start - footer_start..suffix_len - 8])?
+    };
+
+    if enable_page_index {
+        // TODO add async version in arrow-rs avoid read whole file.
```
Comment (Ted-Jiang) [inline datafusion/core/src/datasource/file_format/parquet.rs:422]: will modify in `arrow-rs`

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 2]
```diff
@@ -409,14 +411,31 @@ pub async fn fetch_parquet_metadata(
...
+        let bytes = store.get_range(&meta.location, 0..meta.size).await?;
+        let mut location_vec = vec![];
+        let mut index_vec = vec![];
+        for rg in result_meta.row_groups() {
+            location_vec.push(index_reader::read_pages_locations(&bytes, rg.columns())?);
```
Comment (Ted-Jiang) [inline datafusion/core/src/datasource/file_format/parquet.rs:427]: For now, will read all cols index, Maybe will can modify the `pub async fn fetch_parquet_metadata` API add `index_projection`, if its bottleneck.

**Comment by Ted-Jiang** [comment]
@alamb @liukun4515 @thinkharderdev  PTAL

**Comment by thinkharderdev** [comment]
Thanks @Ted-Jiang. I think it may be better to just pass the option to `ParquetRecordBatchStreamBuilder`. It will fetch the pages indexes in concurrently and if users would like to cache the indexes (or fetch them from somewhere other than the file metadata they can do so with a custom `AsyncFileReader`)

**Comment by Ted-Jiang** [comment]
> ParquetRecordBatchStreamBuilder

Thanks for your advice! @thinkharderdev , IMOP,  if we want to read `page_index` we must get its location from metadata, so it must read after metadata. i think put it directly behind read metadata is better. 

> if users would like to cache the indexes other than the file metadata 
the reason why i put the index in `metadata` is to reduce the code change in repo, if the user want to cache it anywhere else can use pub  API in `page_index::index_reader` to read with metadata.

If i am wrong or misunderstand plz correct me 😊?

**Comment by thinkharderdev** [comment]
> > pass the option to ParquetRecordBatchStreamBuilder
> 
> Thanks for your advice! @thinkharderdev , IMOP, if we want to read `page_index` we must get its location from metadata, so it must read after metadata. i think put it directly behind read metadata is better.
> 
> > if users would like to cache the indexes other than the file metadata
> 
> the reason why i put the index in `metadata` is to reduce the code change in repo, if the user want to cache it anywhere else can use pub API in `page_index::index_reader` to read with metadata.
> 
> If i am wrong or misunderstand plz correct me 😊?

Correct, the advantage of what `ParquetRecordBatchStreamBuilder` does in arrow-rs is that it fill fetch all the indexes concurrently using `AsyncFileReader::get_ranges`. So if you don't read the metadata up front then it will fetch it automatically for you. If we do want to grab it up front in `ParquetExec` then we should copy the implementation in arrow-rs to grab everything concurrently.

**Comment by Ted-Jiang** [comment]
> AsyncFileReader

@thinkharderdev Thanks!  If am right, we should make the read index api into `async`, so i will file a ticket to replace below into async add to arrow-rs to make it concurrently as you mentioned.
```
  // TODO add async version in arrow-rs avoid read whole file.
        let bytes = store.get_range(&meta.location, 0..meta.size).await?;
        let mut location_vec = vec![];
        let mut index_vec = vec![];
        for rg in result_meta.row_groups() {
            location_vec.push(index_reader::read_pages_locations(&bytes, rg.columns())?);
            index_vec.push(index_reader::read_columns_indexes(&bytes, rg.columns())?);
        }

```
I prefer keep read page_index in `ParquetFileReader::get_metadata` and save them in `ParquetMetaData` already define in arrow-rs
```
/// Global Parquet metadata.
#[derive(Debug, Clone)]
pub struct ParquetMetaData {
    file_metadata: FileMetaData,
    row_groups: Vec<RowGroupMetaData>,
    /// Page index for all pages in each column chunk
    page_indexes: Option<ParquetColumnIndex>,
    /// Offset index for all pages in each column chunk
    offset_indexes: Option<ParquetOffsetIndex>,
}
```
because it can reduce the code change ,
Secondly in `parquet open file` first thing we should do is `read file metadata`, following `build_row_filter`, `build_selection_base_on_index`(todo) should depend on this.🤔

**Comment by thinkharderdev** [comment]
> > AsyncFileReader
> 
> @thinkharderdev Thanks! If am right, we should make the read index api into `async`, so i will file a ticket to replace below into async base on `AsyncFileReader::get_ranges` add to arrow-rs to make it concurrently as you mentioned.
> 
> ```
>   // TODO add async version in arrow-rs avoid read whole file.
>         let bytes = store.get_range(&meta.location, 0..meta.size).await?;
>         let mut location_vec = vec![];
>         let mut index_vec = vec![];
>         for rg in result_meta.row_groups() {
>             location_vec.push(index_reader::read_pages_locations(&bytes, rg.columns())?);
>             index_vec.push(index_reader::read_columns_indexes(&bytes, rg.columns())?);
>         }
> ```
> 
> I prefer keep read page_index in `ParquetFileReader::get_metadata` and save them in `ParquetMetaData` already define in arrow-rs
> 
> ```
> /// Global Parquet metadata.
> #[derive(Debug, Clone)]
> pub struct ParquetMetaData {
>     file_metadata: FileMetaData,
>     row_groups: Vec<RowGroupMetaData>,
>     /// Page index for all pages in each column chunk
>     page_indexes: Option<ParquetColumnIndex>,
>     /// Offset index for all pages in each column chunk
>     offset_indexes: Option<ParquetOffsetIndex>,
> }
> ```
> 
> because it can reduce the code change , Secondly in `parquet open file` first thing we should do is `read file metadata`, following `build_row_filter`, `build_selection_base_on_index`(todo) should depend on this.🤔

I think we may be talking about different things :). 

I'm saying the code to fetch the indexes already exists in arrow-rs so we don't need to duplicate the code in datafusion. You can just construct the `ArrowReadOptions`  to enable the page index and `ParquetRecordBatchStreamBuilder` will fetch the indexes (and do so concurrently) (see https://github.com/apache/arrow-rs/blob/a7cf274765945af4111fddaeec26d672715de9d0/parquet/src/arrow/async_reader.rs#L225). 

```
let mut options = ArrowReaderOptions::new().with_page_index(true);

if enable_page_index {
   options = options.with_page_index(true);
}

let builder =
   ParquetRecordBatchStreamBuilder::new_with_options(async_reader, options)
      .await?
```

**Comment by Ted-Jiang** [comment]
> > AsyncFileReader
> 
> @thinkharderdev Thanks! If am right, we should make the read index api into `async`, so i will file a ticket to replace below into async base on `AsyncFileReader::get_ranges` add to arrow-rs to make it concurrently as you mentioned.
> 
> ```
>   // TODO add async version in arrow-rs avoid read whole file.
>         let bytes = store.get_range(&meta.location, 0..meta.size).await?;
>         let mut location_vec = vec![];
>         let mut index_vec = vec![];
>         for rg in result_meta.row_groups() {
>             location_vec.push(index_reader::read_pages_locations(&bytes, rg.columns())?);
>             index_vec.push(index_reader::read_columns_indexes(&bytes, rg.columns())?);
>         }
> ```
> 
> I prefer keep read page_index in `ParquetFileReader::get_metadata` and save them in `ParquetMetaData` already define in arrow-rs
> 
> ```
> /// Global Parquet metadata.
> #[derive(Debug, Clone)]
> pub struct ParquetMetaData {
>     file_metadata: FileMetaData,
>     row_groups: Vec<RowGroupMetaData>,
>     /// Page index for all pages in each column chunk
>     page_indexes: Option<ParquetColumnIndex>,
>     /// Offset index for all pages in each column chunk
>     offset_indexes: Option<ParquetOffsetIndex>,
> }
> ```
> 
> because it can reduce the code change , Secondly in `parquet open file` first thing we should do is `read file metadata`, following `build_row_filter`, `build_selection_base_on_index`(todo) should depend on this.🤔

**Comment by Ted-Jiang** [comment]
@thinkharderdev Sorry for the mistake, i think its ready.

**Comment by tustvold** [comment]
This is modifying submodules, is this intentional?

**Comment by Ted-Jiang** [comment]
> This is modifying submodules, is this intentional?

update the submodules for using the new added test parquet files `alltypes_tiny_pages.parquet` and  `alltypes_tiny_pages_plain.parquet` with page index 
https://github.com/apache/parquet-testing/pull/25

File: datafusion/core/src/physical_plan/file_format/parquet.rs [hunk 3]
```diff
@@ -78,6 +79,10 @@ pub struct ParquetScanOptions {
     /// If true, the generated `RowFilter` may reorder the predicate `Expr`s to try and optimize
     /// the cost of filter evaluation.
     reorder_predicates: bool,
+    /// If true, the reader will read pageIndex, If exit, first we can use it create the `RowSelector`
+    /// before read the file, Second with pageIndex it will accelerate skip records (avoid decode pageHeader)
+    /// when reading values from chunk with `RowSelector`.
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/parquet.rs:84]:
```suggestion
    /// If enabled, the reader will read the page index
    /// This is used to optimise filter pushdown
    /// via `RowSelector` and `RowFilter` by
    /// eliminating unnecessary IO and decoding
```

**Review by tustvold — APPROVED** [review]
Looks good to me, only a minor suggestion.

Thank you for working on this, do you intend to work on hooking up the column index as well?

**Comment by Ted-Jiang** [comment]
> Looks good to me, only a minor suggestion.
> 
> Thank you for working on this, do you intend to work on hooking up the column index as well?

yes,  i plan to do using` min/max stats` and `filter` to create the `RowSelector` 🤔

**Review by alamb — COMMENTED** [review]
This is great -- thank you @Ted-Jiang and @tustvold

**Pull request closed without merging**